docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+The label selector will contain the process class and the cluster name, to ensure processes with the same class are distributed across different fault domains is possible.
+A user can define additional `PodAntiAffinity` but can't prevent the operator from createn the default `PodAntiAffinity`.
+The only expection is the specival fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we would use the process number modulo the `DesiredFaultDomains` and a prefix for the `BinPack` mode.
+In addition to that we have to change the `PodAntiAffinity` rule.
+The following would be an example of the `affinity` term for a Pod.
+The example assumes that process number modulo `DesiredFaultDomains` is `0` and we have set `BinPackAll`:
+
+```yaml
+  affinity:
+    # Affinity to schedule Pods with the same distribution key together
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+              key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across fault domains.
+If `Required` is set to false this will be on a best effort base and can be useful if a fault domain is full but the Pod should be scheduled anyway.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
+    // DesiredFaultDomains desfines the number of desired fault domain.
+    // Should be greater than 0 is fault domain distribution is enabled.
+    // Default: 0
+    DesiredFaultDomains *int
```

Comment:
Why does this default to 0 if the 0 is an invalid value when the bin packing is enabled? Could we set it based on the replication mode by default?

Reply:
We might also want to talk about how this could replace of the special fault domains that we have today, `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`. That overlaps with the question of whether we want to use the logical fault domains for replication in the database, instead of the physical fault domains.

Reply:
Using a different default from 0 based on the replication mode makes sense. I'll rephrase that and add more information. I didn't want to change the FDB zone ID in this design and would leave it to another design to change that. I think about that and if I get a good idea I'll add that here.